File: code-tiny/tools/servlet_jsp/cache.py
from __future__ import annotations

import hashlib
import json
import os
import re
import stat
import tempfile
from typing import Any, Dict, Optional, Tuple

from tools.servlet_jsp.models import SERVLET_JSP_PARSER_VERSION, ResourceBudgets, ServletJspAnalysisResult


SNAPSHOT_SCHEMA_VERSION = 1
_SAFE_SEGMENT_RE = re.compile(r"[^A-Za-z0-9_.-]+")


def servlet_jsp_cache_dir(cache_dir: Optional[str], root: str, project_id: str) -> str:
    if cache_dir:
        base = os.path.realpath(os.path.abspath(os.path.expanduser(cache_dir)))
    else:
        base = os.path.join(
            os.path.realpath(os.path.abspath(os.path.expanduser(os.environ.get("XDG_CACHE_HOME", "~/.cache")))),
            "hyper-graph",
        )
    root_digest = hashlib.sha256(os.path.realpath(root).encode("utf-8")).hexdigest()[:16]
    target = os.path.join(base, "servlet_jsp", f"{_safe_segment(project_id)}-{root_digest}")
    _secure_makedirs(target)
    return target


def preview_artifact_path(cache_dir: Optional[str], root: str, project_id: str) -> str:
    return os.path.join(servlet_jsp_cache_dir(cache_dir, root, project_id), "servlet_jsp_preview.json")


def generation_snapshot_path(
    cache_dir: Optional[str],
    root: str,
    project_id: str,
    module_id: str,
    generation_id: str,
) -> str:
    filename = f"applied-{_safe_segment(module_id)}-{_safe_segment(generation_id)}.json"
    return os.path.join(servlet_jsp_cache_dir(cache_dir, root, project_id), filename)


def write_preview_artifact(path: str, result: ServletJspAnalysisResult) -> str:
    payload = {
        "artifact_role": "preview",
        "schema_version": SNAPSHOT_SCHEMA_VERSION,
        "parser_version": SERVLET_JSP_PARSER_VERSION,
        "result": result.to_dict(),
    }
    return secure_atomic_json_write(path, payload)


def write_generation_snapshot(
    path: str,
    result: ServletJspAnalysisResult,
    *,
    module_id: str,
    generation_id: str,
    budgets: ResourceBudgets,
) -> str:
    return secure_atomic_json_write(
        path,
        generation_snapshot_payload(
            result,
            module_id=module_id,
            generation_id=generation_id,
            budgets=budgets,
        ),
    )


def generation_snapshot_payload(
    result: ServletJspAnalysisResult,
    *,
    module_id: str,
    generation_id: str,
    budgets: ResourceBudgets,
) -> Dict[str, Any]:
    return {
        "artifact_role": "graph_applied_generation",
        "schema_version": SNAPSHOT_SCHEMA_VERSION,
        "parser_version": SERVLET_JSP_PARSER_VERSION,
        "project_id": result.project_id,
        "project_root_digest": hashlib.sha256(os.path.realpath(result.root).encode("utf-8")).hexdigest(),
        "module_id": module_id,
        "generation_id": generation_id,
        "budget_fingerprint": budgets.fingerprint(),
        "result": result.to_dict(),
    }


def generation_snapshot_checksum(
    result: ServletJspAnalysisResult,
    *,
    module_id: str,
    generation_id: str,
    budgets: ResourceBudgets,
) -> str:
    return _payload_checksum(
        generation_snapshot_payload(
            result,
            module_id=module_id,
            generation_id=generation_id,
            budgets=budgets,
        )
    )


def load_generation_snapshot(
    path: str,
    *,
    root: str,
    project_id: str,
    module_id: str,
    generation_id: str,
    expected_checksum: str,
    budgets: ResourceBudgets,
    max_bytes: int = 512 * 1024 * 1024,
) -> Tuple[Optional[Dict[str, Any]], str]:
    try:
        info = os.lstat(path)
        if stat.S_ISLNK(info.st_mode) or not stat.S_ISREG(info.st_mode):
            return None, "snapshot is not a regular non-symlink file"
        if info.st_size > max_bytes:
            return None, "snapshot exceeds the configured size limit"
        with open(path, "rb") as handle:
            raw = handle.read(max_bytes + 1)
        if len(raw) > max_bytes:
            return None, "snapshot exceeds the configured size limit"
        envelope = json.loads(raw.decode("utf-8"))
    except (OSError, UnicodeError, json.JSONDecodeError) as exc:
        return None, f"snapshot is unreadable: {exc}"
    checksum = str(envelope.pop("payload_sha256", ""))
    actual = _payload_checksum(envelope)
    if not checksum or checksum != actual or (expected_checksum and expected_checksum != actual):
        return None, "snapshot checksum mismatch"
    expected = {
        "artifact_role": "graph_applied_generation",
        "schema_version": SNAPSHOT_SCHEMA_VERSION,
        "parser_version": SERVLET_JSP_PARSER_VERSION,
        "project_id": project_id,
        "project_root_digest": hashlib.sha256(os.path.realpath(root).encode("utf-8")).hexdigest(),
        "module_id": module_id,
        "generation_id": generation_id,
        "budget_fingerprint": budgets.fingerprint(),
    }
    for key, value in expected.items():
        if envelope.get(key) != value:
            return None, f"snapshot metadata mismatch: {key}"
    return envelope, "ok"


def secure_atomic_json_write(path: str, payload: Dict[str, Any]) -> str:
    destination = os.path.abspath(path)
    parent = os.path.dirname(destination)
    _secure_makedirs(parent)
    if os.path.lexists(destination) and stat.S_ISLNK(os.lstat(destination).st_mode):
        raise OSError(f"Refusing to replace symlinked output: {destination}")
    envelope = dict(payload)
    checksum = _payload_checksum(envelope)
    envelope["payload_sha256"] = checksum
    fd, temp_path = tempfile.mkstemp(prefix=f".{os.path.basename(destination)}.", suffix=".tmp", dir=parent, text=True)
    try:
        os.fchmod(fd, 0o600)
        with os.fdopen(fd, "w", encoding="utf-8") as handle:
            json.dump(envelope, handle, ensure_ascii=True, indent=2, sort_keys=True)
            handle.write("\n")
            handle.flush()
            os.fsync(handle.fileno())
        os.replace(temp_path, destination)
        dir_fd = os.open(parent, os.O_RDONLY)
        try:
            os.fsync(dir_fd)
        finally:
            os.close(dir_fd)
    except Exception:
        try:
            os.unlink(temp_path)
        except OSError:
            pass
        raise
    return checksum


def _payload_checksum(payload: Dict[str, Any]) -> str:
    encoded = json.dumps(payload, ensure_ascii=True, sort_keys=True, separators=(",", ":")).encode("utf-8")
    return hashlib.sha256(encoded).hexdigest()


def _secure_makedirs(path: str) -> None:
    absolute = os.path.abspath(path)
    drive, tail = os.path.splitdrive(absolute)
    current = drive + os.sep if absolute.startswith(os.sep) else drive
    for part in [item for item in tail.split(os.sep) if item]:
        current = os.path.join(current, part) if current else part
        if os.path.lexists(current):
            info = os.lstat(current)
            if stat.S_ISLNK(info.st_mode) or not stat.S_ISDIR(info.st_mode):
                raise OSError(f"Unsafe cache/output path component: {current}")
            continue
        os.mkdir(current, 0o700)
    try:
        os.chmod(absolute, 0o700)
    except OSError:
        pass


def _safe_segment(value: str) -> str:
    return _SAFE_SEGMENT_RE.sub("_", (value or "").strip()).strip("._") or "value"

File: rust/crates/analyzer-jvm-overlays/src/servlet_jsp/cache.rs
//! Port `tools/servlet_jsp/cache.py` — preview/snapshot artifacts (checksum
//! envelope `payload_sha256`).

use std::path::{Path, PathBuf};

use serde_json::{json, Value};

use crate::servlet_jsp::models::{stable_digest, ResourceBudgets, ServletJspAnalysisResult, SERVLET_JSP_PARSER_VERSION};
use crate::pyutil::{realpath, sha256_hex};

pub const SNAPSHOT_SCHEMA_VERSION: i64 = 1;

/// `_safe_segment`.
pub fn safe_segment(value: &str) -> String {
    static RE: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
    let re = RE.get_or_init(|| regex::Regex::new(r"[^A-Za-z0-9_.-]+").unwrap());
    let replaced = re.replace_all(value.trim(), "_").to_string();
    let cleaned = replaced.trim_matches(|c| c == '.' || c == '_');
    if cleaned.is_empty() {
        "value".to_string()
    } else {
        cleaned.to_string()
    }
}

/// `servlet_jsp_cache_dir`.
pub fn servlet_jsp_cache_dir(cache_dir: Option<&str>, root: &str, project_id: &str) -> PathBuf {
    let base = match cache_dir {
        Some(dir) if !dir.is_empty() => crate::pyutil::realpath(Path::new(dir)).to_string_lossy().to_string(),
        _ => {
            let xdg = std::env::var("XDG_CACHE_HOME").unwrap_or_else(|_| "~/.cache".to_string());
            let expanded = if let Some(rest) = xdg.strip_prefix("~/") {
                std::env::var("HOME").map(|home| format!("{home}/{rest}")).unwrap_or(xdg)
            } else {
                xdg
            };
            crate::pyutil::realpath(Path::new(&expanded))
                .join("hyper-graph")
                .to_string_lossy()
                .to_string()
        }
    };
    let root_digest = sha256_hex(realpath(Path::new(root)).to_string_lossy().as_bytes())[..16].to_string();
    let target = PathBuf::from(base)
        .join("servlet_jsp")
        .join(format!("{}-{}", safe_segment(project_id), root_digest));
    let _ = std::fs::create_dir_all(&target);
    target
}

pub fn preview_artifact_path(cache_dir: Option<&str>, root: &str, project_id: &str) -> PathBuf {
    servlet_jsp_cache_dir(cache_dir, root, project_id).join("servlet_jsp_preview.json")
}

pub fn generation_snapshot_path(
    cache_dir: Option<&str>,
    root: &str,
    project_id: &str,
    module_id: &str,
    generation_id: &str,
) -> PathBuf {
    let filename = format!("applied-{}-{}.json", safe_segment(module_id), safe_segment(generation_id));
    servlet_jsp_cache_dir(cache_dir, root, project_id).join(filename)
}

/// `_payload_checksum`.
pub fn payload_checksum(payload: &Value) -> String {
    sha256_hex(crate::pyjson::dumps_compact(payload).as_bytes())
}

pub fn generation_snapshot_payload(
    result: &ServletJspAnalysisResult,
    module_id: &str,
    generation_id: &str,
    budgets: &ResourceBudgets,
) -> Value {
    crate::pyjson::py_object(vec![
        ("artifact_role".into(), json!("graph_applied_generation")),
        ("schema_version".into(), json!(SNAPSHOT_SCHEMA_VERSION)),
        ("parser_version".into(), json!(SERVLET_JSP_PARSER_VERSION)),
        ("project_id".into(), json!(result.project_id)),
        (
            "project_root_digest".into(),
            json!(sha256_hex(realpath(Path::new(&result.root)).to_string_lossy().as_bytes())),
        ),
        ("module_id".into(), json!(module_id)),
        ("generation_id".into(), json!(generation_id)),
        ("budget_fingerprint".into(), json!(budgets.fingerprint())),
        ("result".into(), result.to_dict()),
    ])
}

pub fn generation_snapshot_checksum(
    result: &ServletJspAnalysisResult,
    module_id: &str,
    generation_id: &str,
    budgets: &ResourceBudgets,
) -> String {
    payload_checksum(&generation_snapshot_payload(result, module_id, generation_id, budgets))
}

/// `write_preview_artifact`.
pub fn write_preview_artifact(path: &Path, result: &ServletJspAnalysisResult) -> Result<String, String> {
    let payload = crate::pyjson::py_object(vec![
        ("artifact_role".into(), json!("preview")),
        ("schema_version".into(), json!(SNAPSHOT_SCHEMA_VERSION)),
        ("parser_version".into(), json!(SERVLET_JSP_PARSER_VERSION)),
        ("result".into(), result.to_dict()),
    ]);
    secure_atomic_json_write(path, &payload)
}

/// `write_generation_snapshot`.
pub fn write_generation_snapshot(
    path: &Path,
    result: &ServletJspAnalysisResult,
    module_id: &str,
    generation_id: &str,
    budgets: &ResourceBudgets,
) -> Result<String, String> {
    secure_atomic_json_write(
        path,
        &generation_snapshot_payload(result, module_id, generation_id, budgets),
    )
}

/// `secure_atomic_json_write` — envelope + checksum, indent 2 sort_keys
/// ensure_ascii + "\n", chmod 600.
pub fn secure_atomic_json_write(path: &Path, payload: &Value) -> Result<String, String> {
    let destination = crate::pyutil::abs_path(path);
    let parent = destination
        .parent()
        .map(|parent| parent.to_path_buf())
        .unwrap_or_else(|| PathBuf::from("."));
    std::fs::create_dir_all(&parent).map_err(|error| error.to_string())?;
    if destination.exists() {
        let metadata = std::fs::symlink_metadata(&destination).map_err(|error| error.to_string())?;
        if metadata.file_type().is_symlink() {
            return Err(format!("Refusing to replace symlinked output: {}", destination.display()));
        }
    }
    let checksum = payload_checksum(payload);
    let mut envelope = payload.as_object().cloned().unwrap_or_default();
    envelope.insert("payload_sha256".into(), Value::String(checksum.clone()));
    let body = crate::pyjson::dumps_pretty(&Value::Object(envelope));
    let tmp_path = parent.join(format!(
        ".{}.{}.tmp",
        destination.file_name().map(|n| n.to_string_lossy()).unwrap_or_default(),
        stable_digest(&[format!("{:?}", std::time::SystemTime::now())], 12)
    ));
    std::fs::write(&tmp_path, body).map_err(|error| error.to_string())?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let _ = std::fs::set_permissions(&tmp_path, std::fs::Permissions::from_mode(0o600));
    }
    std::fs::rename(&tmp_path, &destination).map_err(|error| error.to_string())?;
    Ok(checksum)
}
